src/sm/networkmanager/networkmanager.cpp

```diff
@@ -0,0 +1,747 @@
+/*
+ * Copyright (C) 2024 EPAM Systems, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "aos/sm/networkmanager.hpp"
+
+#include "log.hpp"
+
+namespace aos {
+namespace sm {
+namespace networkmanager {
+
+/***********************************************************************************************************************
+ * Public
+ **********************************************************************************************************************/
+
+Error NetworkManager::Init(StorageItf& storage, CNIItf& cni, TrafficMonitorItf& netMonitor, const String& workingDir)
+{
+    LOG_DBG() << "Initialize network manager";
+
+    mStorage    = &storage;
+    mCNI        = &cni;
+    mNetMonitor = &netMonitor;
+
+    auto cniDir = FS::JoinPath(workingDir, "cni");
+
+    if (auto err = mCNI->Init(cniDir); !err.IsNone()) {
+        return AOS_ERROR_WRAP(err);
+    }
+
+    mCNINetworkCacheDir = FS::JoinPath(cniDir, "networks");
+
+    return AOS_ERROR_WRAP(mNetMonitor->Start());
+}
+
+Error NetworkManager::Close()
+{
+    LOG_DBG() << "Close network manager";
+
+    return AOS_ERROR_WRAP(mNetMonitor->Close());
+}
+
+Error NetworkManager::UpdateNetworks(const Array<aos::NetworkParameters>& networks)
+{
+    LOG_DBG() << "Update networks";
+
+    (void)networks;
+
+    return ErrorEnum::eNone;
+}
+
+Error NetworkManager::AddInstanceToNetwork(
+    const String& instanceID, const String& networkID, const NetworkParams& network)
+{
+    LOG_DBG() << "Add instance to network: instanceID=" << instanceID << ", networkID=" << networkID;
+
+    auto err = IsInstanceInNetwork(instanceID, networkID);
+    if (err.IsNone()) {
+        return ErrorEnum::eAlreadyExist;
+    }
+
+    if (!err.Is(ErrorEnum::eNotFound)) {
+        return err;
+    }
+
+    if (err = AddInstanceToCache(instanceID, networkID); !err.IsNone()) {
+        return err;
+    }
+
+    auto cleanupInstanceCache = [this, instanceID, networkID]() {
```

Comment:
We have custom deleter for unique ptr, could it be used here?

Reply:
Done